release-23.1: ccl: add an option to perform a regionless RESTORE

annrpom opened the pull request:
Backport:
  * 1/1 commits from "ccl: add an option to perform a regionless RESTORE" (#110606)
  * 1/1 commits from "ccl: syntax change restore option strip_localities to remove_regions" (#111356)
  * 1/2 commits from "ccl: regionless restores with rbr table(s) fail fast, add version gate" (#111443)

Please see individual PRs for details.

---
### ccl: add cluster setting for remove_regions option

This patch adds a cluster setting, `bulkio.restore.remove_regions.enabled`,
that is disabled (set to false) by default - restricting users to use
remove_regions unless this setting is toggled.

Epic: none

Release note: None

--

/cc @cockroachdb/release

Release justification: This is a high-priority, opt-in change that has all code affected guarded by a check.

annrpom committed:
ccl: add an option to perform a regionless RESTORE

This patch adds the ability to restore a multi-region database onto a
single region cluster by adding a new option for the RESTORE syntax,
`strip_localities`. This option will strip RegionConfigs from
databases and LocalityConfigs from tables.

Epic: none
Fixes: #105623
Release note (sql change): A new option for the RESTORE
syntax, `strip_localities`, has been added. This can be used to
strip the locality information from a backup when there
are mismatched cluster regions between the backup's
cluster and the target cluster.

The following are behaviors that will most likely not be encountered
with the specific use case that this patch provides, but are
documented nonetheless:

Note: Adding a primary region to a regionless restore (with or
without regional by row table(s)) will not work out-of-the-box, but
does produce an accurate message detailing a user to essentially
`DROP TYPE [database].public.crdb_internal_region;` (and for cluster
restores, `ALTER DATABASE [database] CONFIGURE ZONE DISCARD;`).

Note: Restoring a cluster/database/table with a regional by row table
will not work out-of-the box. In particular, if we are performing
writes, the `crdb_region` column needs to specify the region of the
new row(s) being written to the table. The user will need to alter
said column and set a default that makes sense, along with discarding
the zone config (this latter is due to the fact that the zone config
holds all outdated info related to the partitions, constraints,
etc.).

These are due to a conflict with the `crdb_region` column already
being present in the regionless restore.  This column specifies each
row's home region and is a prefix to the table's primary key.
Stripping localities does not touch this column as it would be an
expensive operation that includes rewriting the entire table
([internal discussion here](https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1694618761011619)).

annrpom committed:
ccl: syntax change restore option strip_localities to remove_regions

This changes the syntax from newly added RESTORE option,
strip_localities, to remove_regions. The former was
inadequate because it was too similar to an existing
option, along with not being an accurately descriptive enough
name.

Epic: none
Informs: #111348
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-29129

Release note (sql change): the syntax for RESTORE option,
strip_localities, is now remove_regions.

annrpom committed:
ccl: regionless restores with regional by row table(s) fails fast

Previously, backing up an MR cluster/db/table that had a regional by
row table did not work out-of-the-box (users could not write to said
table without performing some operations to ensure hidden column added
by regional by row enablement, `crdb_region`, and zone config behaved
according to the target RESTORE cluster. This was inadequate because
we allowed users to perform a RESTORE where some tables were not
"available" right away. This patch addresses this by making sure we
"fail fast" if users do decide to attempt a RESTORE where the BACKUP
object has a regional by row table.

Also: tests. A small amount of tests were fixed to separate the behavior
of restoring a database vs a table (just dropping a table after we
performed a db restore to test table restore is not a properly
isolated test of behavior).

Epic: none
Informs: #111348
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-29129

Release note (sql change): Users are not able to perform a
`remove_regions` RESTORE if the object being restored contains
a regional by row table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111863)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1067,6 +1068,12 @@ func restorePlanHook(
 			errors.New("cannot run RESTORE with schema_only until cluster has fully upgraded to 22.2")
 	}
+
+	if restoreStmt.Options.RemoveRegions && !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.V23_1) {
```
I don't think this is doing what we think it is doing here: this check can return true -- that is, V23_1 is active -- even if there are nodes in the cluster that do not contain this change. That is, there is no way, in a backport, to be sure before you write or send something that the node that reads/receives it will be running your new code -- it could still be running v23.1.9 or something, but IsActive(V23_1) is true.

rafiss replied:
in a team discussion, we were thinking of introducing a new hidden cluster setting in the backport, which is named something like `restore.remove_regions.enabled`. the new restore option would only be allowed if this setting is on, and if a customer sets it, we'll know they only did so if we advised them to do so and it can serve as implicit confirmation that all nodes are on the new 23.1 patch release.

annrpom committed:
ccl: add cluster setting for remove_regions option

This patch adds a cluster setting, `bulkio.restore.remove_regions.enabled`,
that is disabled (set to false) by default - restricting users to use
remove_regions unless this setting is toggled.

Epic: none

Release note: None

annrpom commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -97,6 +97,13 @@ var featureRestoreEnabled = settings.RegisterBoolSetting(
 	featureflag.FeatureFlagEnabledDefault,
 ).WithPublic()
 
+var optionRemoveRegionsEnabled = settings.RegisterBoolSetting(
```
thanks for pointing this out - qq (might be a silly q): would it be worth specifying the exact patch in the error message we produce if this cluster setting is disabled? currently, it just prompts the user to ensure all nodes are "on the same patch of 23.1". not sure if there would be a case where someone could interpret that to mean a patch that does not have this change - resulting in some syntax error with remove_regions

rafiss replied:
that's a good point. i think it should say
```
errors.Newf("to use the remove_regions option, ensure that all nodes are running version %s " +
	"or greater; then, enable cluster setting bulkio.restore.remove_regions.enabled", build.GetInfo().Tag)`
```

rafiss reviewed (approved):
this lgtm! can you post a backport request in #db-backports-point-releases ?

The policy is: https://cockroachlabs.atlassian.net/wiki/x/_YEuPg

dikshant reviewed (approved):
This was needed for a customer who is upgrading to 23.1 in Nov. The business/customer need is referenced in this thread: https://cockroachlabs.slack.com/archives/C2C5FKPPB/p1695919462529609?thread_ts=1695916743.657659&cid=C2C5FKPPB

dt commented:
I am 👍  that the new functionality here insofar as it affects DR's portions of backup/restore is reachable only via opt-in (by using the new flag, the new fields being populated in the job by using the new flag). 

I'm not quite sure the setting gate in front of the flag is doing much for us? As far as an opt-in, we already had that just by passing the option and as far as mixed version protection, there is nothing keeping you from setting it when you still have old nodes any more than there is keeping you from passing the new option when you have old nodes, but apparently this is what we've done in the past. As long as it is non-public and we don't document it, then we can use the conversation where we mention the setting to also drill home the importance of verifying the mixed version state is not a concern (if we add this setting in this backport however, we do need to go retire it on master even if it was never in master).

rafiss commented:
Backport review:

> Are all of the changes protected via a flag or option, new syntax, an environment variable or default-disabled cluster setting?

Yes, there is an off-by-default cluster setting. The new RESTORE option can only be used with that on.

> What are the risks to backporting this change?  Can the risks of backporting be mitigated?  What are the risks to not backporting?   

The risk of backporting is that a user might unintentionally remove regions when restoring. This is mitigated since that behavior is opt-in and requires a cluster setting to be enabled. The risk to not backporting is that a user would be unable to restore a multiregion backup in a single-node cluster, which is sometimes desired for testing purposes. `skip_missing_localities` is another option that used to help with this, but it no longer is sufficient to rely on that alone in 23.1.

> Does this change really need to be backported? Or can it reasonably wait until the next major release to be addressed?

See @dikshant's comment above.
